ouroboros-consensus/src/Ouroboros/Consensus/HardFork/History/Follower.hs

```diff
@@ -0,0 +1,144 @@
+    }
+
+data WithInterpreter blk b = WithInterpreter {
+      payload     :: b
+    , interpreter :: Maybe (Interpreter (HardForkIndices blk))
+    } deriving Show
+
+type WithInterDefault blk = WithInterpreter blk (Serialised blk)
+
+mkWithInterpreter :: b -> Maybe (Interpreter (HardForkIndices blk)) -> WithInterpreter blk b
+mkWithInterpreter b mInter = WithInterpreter {
+      payload     = b
+    , interpreter = mInter
+    }
+
+instance (ShowProxy blk, ShowProxy b) => ShowProxy (WithInterpreter blk b) where
+    showProxy _ = mconcat [
+        "WithInterpreter "
+      , showProxy (Proxy :: Proxy blk)
+      , " "
+      , showProxy (Proxy :: Proxy b)
+      ]
+
+{-------------------------------------------------------------------------------
+  Implementation
+-------------------------------------------------------------------------------}
+
+newHistoryFollower ::
+    forall blk m.
+    ( IOLike m
+    , HasHardForkHistory blk
+    , IsLedger (LedgerState blk)
+    )
+    => LedgerConfig blk
+    -> ChainDB m blk
+    -> m (HistoryFollower m blk)
+newHistoryFollower cfg chainDB = do
+    varState <- newTVarIO Nothing
+    return $ HistoryFollower {
+        getInterpreter =
+          getInterpreterImpl cfg (ledgerState <$> getCurrentLedger chainDB) varState
+      }
+
+getInterpreterImpl ::
+       (IOLike m, HasHardForkHistory blk)
+    => LedgerConfig blk
+    -> STM m (LedgerState blk)
+    -> StrictTVar m (Maybe (Interpreter (HardForkIndices blk)))
+    -> Point blk
+    -> STM m (Maybe (Interpreter (HardForkIndices blk)))
+getInterpreterImpl cfg getState varInterpreter point = do
+    minter <- readTVar varInterpreter
+    case (minter, pointSlot point) of
+      (Just inter, At slot) | isSlotInHorizon slot inter -> return Nothing
+      _ -> do
+        st <- getState
+        let inter = History.mkInterpreter $ hardForkSummary cfg st
+        writeTVar varInterpreter $ Just inter
+        return $ Just inter
+
+-- | Run a sanity query to check if the slot is still in the Horizon.
+isSlotInHorizon :: SlotNo -> Interpreter xs -> Bool
+isSlotInHorizon slotNo inter = isRight $ interpretQuery inter query
+  where
+    query = (,) <$> slotToWallclock slotNo <*> slotToEpoch' slotNo
+
+{-------------------------------------------------------------------------------
+  Client History Follower
+-------------------------------------------------------------------------------}
+
+-- | This can be used by a client. Given a 'WithInterClient',
+-- 'getInterpreter' is the best effort to retrieve an 'Interpreter'.
+newtype ClientHistoryFollower m blk = ClientHistoryFollower {
+      getInterpreterClient :: WithInterClient blk
+                           -> STM m (Maybe (Interpreter (HardForkIndices blk)))
+    }
+
+newClientFollower :: MonadSTM m => m (ClientHistoryFollower m blk)
+newClientFollower = do
+    mInter <- newTVarIO Nothing
+    return $ ClientHistoryFollower $ getInterClientImpl mInter
+
+getInterClientImpl :: MonadSTM m
```

Comment:
Please indent the type on a newline after `::`, like you did in `newHistoryFollower`.